autogen/agentchat/contrib/teachable_agent.py

```diff
@@ -0,0 +1,412 @@
+
+
+class TeachableAgent(ConversableAgent):
+    """Teachable Agent, a subclass of ConversableAgent using a vector database to remember user teachings."""
+
+    def __init__(
+        self,
+        name="Agent",  # default set to Assistant
+        system_message: Optional[
+            str
+        ] = "You are a helpful AI assistant that remembers user teachings from prior chats.",
+        llm_config: Optional[Union[Dict, bool]] = None,
+        is_termination_msg: Optional[Callable[[Dict], bool]] = None,
+        max_consecutive_auto_reply: Optional[int] = None,
+        human_input_mode: Optional[str] = "NEVER",
+        code_execution_config: Optional[Union[Dict, bool]] = False,
+        teach_config: Optional[Dict] = None,  # config for the TeachableAgent
+        **kwargs,
+    ):
+        """
+        Args:
+            name (str): name of the agent. Default "Assistant".
+            human_input_mode (str): NEVER ask for human input for this agent.
+            teach_config (dict or None): config for the TeachableAgent.
+                To use default config, set to None. Otherwise, set to a dictionary with any of the following keys:
+                - verbosity (Optional, int): # 0 (default) for basic info, 1 to add memory operations, 2 for analyzer messages, 3 for memo lists.
+                - reset_db (Optional, bool): True to clear the DB before starting. Default False.
+                - path_to_db_dir (Optional, str): path to the directory where the DB is stored. Default "./tmp/teachable_agent_db"
+                - prepopulate (Optional, int): True (default) to prepopulate the DB with a set of input-output pairs.
+                - use_cache (Optional, bool): True to skip LLM calls made previously by relying on cached responses. Default False.
+                - recall_threshold (Optional, float): The distance threshold for retrieving memos from the DB. Default 1.5.
+                - max_num_retrievals (Optional, int): The maximum number of memos to retrieve from the DB. Default 10.
+            **kwargs (dict): other kwargs in [ConversableAgent](../conversable_agent#__init__).
+        """
+        super().__init__(
+            name,
+            system_message,
+            is_termination_msg,
+            max_consecutive_auto_reply,
+            human_input_mode,
+            code_execution_config=code_execution_config,
+            llm_config=llm_config,
+            **kwargs,
+        )
+        self.register_reply(Agent, TeachableAgent._generate_teachable_assistant_reply)
+
+        self._teach_config = {} if teach_config is None else teach_config
+        self.verbosity = self._teach_config.get("verbosity", 0)
+        self.reset_db = self._teach_config.get("reset_db", False)
+        self.path_to_db_dir = self._teach_config.get("path_to_db_dir", "./tmp/teachable_agent_db")
+        self.prepopulate = self._teach_config.get("prepopulate", True)
+        self.use_cache = self._teach_config.get("use_cache", False)
```

Comment:
Why not set default to True to better save costs?

Reply:
It's a good question. I had three main scenarios in mind:
1. Development and testing, where I would set use_cache=False to see all the effects of non-deterministic LLM generation.
2. Demos for management, where I would set use_cache=True to make the demo run fast and avoid surprises.
3. Production use by customers, who may set it either way depending on their needs.

I've usually been in situation 1, and I assume the same applies to most users of this repo. When starting to write and test this agent, I didn't realize that caching was on by default, which made me waste some time thinking that everything was more stable than it really was.

Reply:
Nice trick!